HARK/ConsumptionSaving/ConsIndShockModel_AgentSolve.py

```diff
@@ -0,0 +1,2377 @@
+        object.
+
+        """
+        # Working backwards from End of Problem
+
+        # Later we can select among the allowed transitions
+        # First, for model with shocks at Beginning of Problem/Period (BOP):
+        chosen_to_next_BOP = \
+            {'kNrm': 'kNrm = aNrm'}  # k_{t+1} = a_{t}
+
+        choice_to_chosen = \
+            {'aNrm': 'aNrm = mNrm - cNrm'}  # a_{t} = m_{t} - c_{t}
+
+        BOP_to_choice = {
+            'RNrm': 'RNrm = Rfree / (PermGroFac * permShk)',
+            'bNrm': 'bNrm = kNrm * RNrm',
+            'yNrm': 'yNrm = tranShk',
+            'mNrm': 'mNrm = bNrm + yNrm'}
+
+        # Now, for model with shocks at End of Problem/Period (EOP)
+
+        chosen_to_next_choice = \
+            {'kNrm': 'kNrm = aNrm',
+             'RNrm': 'RNrm = Rfree / (PermGroFac * permShk)',
+             'bNrm': 'bNrm = kNrm * RNrm',  # b_{t} = k_{t} RNrm_{t}
+             'yNrm': 'yNrm = tranShk',  # y_{t} = \tranShk_{t}
+             'mNrm': 'mNrm = bNrm + yNrm'}
+
+        # choice_to_chosen need not be redefined b/c same as defined above
+
+        possible_transitions = \
+            {'chosen_to_next_BOP': chosen_to_next_BOP,
+             'chosen_to_next_choice': chosen_to_next_choice,
+             'choice_to_chosen': choice_to_chosen,
+             'BOP_to_choice': BOP_to_choice
+             }
+
+        return possible_transitions
+
+    def describe_model_and_calibration(self, messaging_level=logging.INFO,
+                                       quietly=False):
+        """
+        Log a brief description of the model and its calibration.
+
+        Parameters
+        ----------
+        self : agent_stage_solution
+
+            Solution to the problem described by information for the current
+        stage found in Bilt and the succeeding stage.
+
+        quietly : boolean, optional
+
+            If true, suppresses all output
+
+        Returns
+        -------
+        None
+        """
+
+        crnt = self
+        Pars, Modl = crnt.Pars, crnt.Modl
+        Tran = Modl.Transitions
+
+        if not quietly and messaging_level < logging.WARNING:
+            msg = '\n(quietly=False and messaging_level < logging.WARNING, ' \
+                + 'so some model information is provided below):\n'
+            msg = msg + '\nThe model has the following parameter values:\n'
+            _log.setLevel(messaging_level)
+            _log.info(msg)
+            for key in Pars.__dict__.keys():
+                _log.info('\t' + key + ': ' + str(Pars.__dict__[key]))
+
+            msg = "\nThe model's transition equations are:"
+            _log.info(msg)
+            for key in Tran.keys():
+                _log.info('\n' + key + ' step:')
+                for eqn_name in Tran[key]['raw_text']:
+                    _log.info('\t' + str(Tran[key]['raw_text'][eqn_name]))
+
+    def check_conditions(self, messaging_level=logging.DEBUG, quietly=False):
+        """
```

Comment:
I wonder if the check_conditions functionality is properly part of the solver functionality or part of the type defintion